### ISSUE TITLE
[Weird Bug On Forwarding Request to Cluster Node]

### ISSUE THREAD

**Issue by riandyrn** [tinode/chat#27]
Hello, Gene

I’m facing a weird bug when trying to set private field of user’s fnd topic which apparently located on different cluster node. It always return following error message: 

```
cluster: forwarding request to node 'two’
cluster: call failed to 'two' [gob: type not registered for interface: []interface {}]
```

The culprit is this line: https://github.com/tinode/chat/blob/devel/server/cluster.go#L135. But I don’t understand why it throws error while for forwarding normal message it works.

Maybe you have idea why this error happens, Gene?

Thanks

**Comment by or-else** [comment]
I'll check it out today.

**Comment by or-else** [comment]
Hi Riandy,
I made a quick fix and pushed it to devel https://github.com/tinode/chat/commit/25ac06e45c6b4256c24583ee014cdf8cb8b477bf.  

That's not the recommended way though. A correct fix would be to replace all interface{} types in https://github.com/tinode/chat/blob/master/server/datamodel.go with a concrete type like json.RawMessage. Unfortunately it will likely affect the data already stored in the database. If you are OK with that I'll try the RawMessage fix.

**Comment by riandyrn** [comment]
> I made a quick fix and pushed it to devel 25ac06e.

Hoo…, I see, so the type need to be registered before the creation of ClusterNode. Previously I was also tried to use `gob.Register()` but instead put it on `clusterInit()`, I put it on `(n *ClusterNode) call()`, so it always return following error: `[gob: name not registered for interface: "[]interface {}”]`. Haha…

> with a concrete type like json.RawMessage

Hmm…, json.RawMessage basically is []byte isn’t it? So if it is saved on database it will be saved as binary, right?

If so, I personally think it’s better to go with current fix, because it is much easier to debug + easier to integrate with another system.

**Comment by or-else** [comment]
> so the type need to be registered

Yes, but not all types, just the interface types.

> json.RawMessage basically is []byte isn’t it? So if it is saved on database it will be saved as binary, right?

Yes, it is and yes it will be saved as binary. 

OK, let's keep this fix for now. We can revisit it if new problems appear.
